### DISCUSSION TITLE
Add range-limited betweenness centrality

### DISCUSSION THREAD

**Discussion by nahuelalmeira** [networkit/networkit#1294] [Q&A]
Hi all,

I am currently doing research related to range-limited betweenness centrality, which consists in computing the betweenness of each node taking only paths up to a given max length. As this feature is not available in Networkit, I am currently using Igraph, which includes it as an argument in the method `betweenness` [1]. Unfortunately, Igraph is not prepared for parallelization as Networkit and thus cannot be used in large networks. It would be great if this feature could be incorporated. As far as I know, it should not imply large changes in code, as the only difference with standard betweenness is that a cutoff is used to stop the breadth first search.

In addition, an efficient algorithm exists where range-limited betweenness can be computed for all cutoffs simultaneously, with no increasing time complexity (see ref. [2-3]). I would be very grateful if this algorithm could be also incorporated.

References:

[1] https://igraph.org/python/doc/api/igraph._igraph.GraphBase.html#betweenness
[2] M. Ercsey-Ravasz and Z. Toroczkai, Centrality scaling in large networks, Phys. Rev. Letters 105, 038701 (2010).
[3] M. Ercsey-Ravasz, et al., Range-limited centrality measures in complex networks, Phys. Rev. E 85, 066103 (2012).

**Comment by hmeyerhenke** [comment]
Dear @nahuelalmeira, would you be interested in developing the code yourself, with some guidance from the maintenance team?

**Comment by nahuelalmeira** [comment]
Hi @hmeyerhenke. I am not an experienced developer, but I think that with your guidance I could give it a try. Could you tell me how to proceed?

**Comment by hmeyerhenke** [comment]
@nahuelalmeira: Please get in touch with @fabratu and @angriman. They can tell you much better how to proceed. Fabian, Eugenio, can you please provide the usual pointers in the docs for new developers?

**Comment by angriman** [comment]
@nahuelalmeira the [NetworKit Development Guide](https://networkit.github.io/dev-docs/DevGuide.html#devGuide) provides all the information you need to know about how to develop and send contributions to NetworKit. Then a possible starting point would be to have a look at how Betweenness is computed (see the implementation [here](https://github.com/networkit/networkit/blob/master/networkit/cpp/centrality/Betweenness.cpp)). For more about NetworKit C++ APIs you can always take a look at the [documentation](https://networkit.github.io/dev-docs/cpp_api/library_root.html).

**Comment by fabratu** [comment]
@nahuelalmeira Did you find the time to make any progress so far?

As an entry-point to begin with - your initial conclusion about BFS is correct. Basicallly a first step would be to add an additional (default initialized) parameter like `count cutoff = 0` to the [constructor of BFS](https://github.com/networkit/networkit/blob/master/include/networkit/distance/BFS.hpp#L34). If you look at the code of our current implementation of the [exact Betweenness](https://github.com/networkit/networkit/blob/master/networkit/cpp/centrality/Betweenness.cpp#L39) (basically Brandes algorithm), you see that the same has to be done for Dijkstra (in case the graph is weighted). In the end you also have to add the cutoff-parameter to the constructor of [Betweenness](https://github.com/networkit/networkit/blob/master/networkit/cpp/centrality/Betweenness.cpp#L19) and pass it during creation of BFS/Dijkstra objects.

As for the general setup of a dev-environment - don't hesitate to ask if there are questions.

Of course implementing [2] and [3] or having a more sophisticated approach would be more complex - but the parallelization should give you a decent speedup (depending on your computing resources).
